File: rs/src/nmea_generator.rs
use std::char::MAX;

use chrono::prelude::*;
use rand::{
    distributions::{Distribution, Uniform},
    rngs::ThreadRng,
    thread_rng, Rng,
};

pub struct RandomGenerator {
    rng: ThreadRng,
}

impl RandomGenerator {
    pub fn new() -> Self {
        RandomGenerator { rng: thread_rng() }
    }

    pub fn random_uniform(&mut self, min: f64, max: f64) -> f64 {
        let range = Uniform::from(min..max);
        range.sample(&mut self.rng)
    }

    pub fn random_int(&mut self, min: i32, max: i32) -> i32 {
        let range = Uniform::from(min..max);
        range.sample(&mut self.rng)
    }
}

#[derive(Debug, Clone)]
struct Satellite {
    constellation: Constellation,
    id: u16,
}

impl Satellite {
    pub fn new(constell: Constellation, id: u16) -> Self {
        Satellite {
            constellation: constell,
            id,
        }
    }

    pub fn new_random() -> Self {
        let constell = Constellation::get_random();
        let id = match constell {
            Constellation::GPS => RandomGenerator::new().random_int(1, 32),
            Constellation::GLONASS => RandomGenerator::new().random_int(65, 96),
            Constellation::GALILEO => RandomGenerator::new().random_int(1, 36),
            Constellation::BEIDOU => RandomGenerator::new().random_int(101, 136),
            Constellation::QZSS => RandomGenerator::new().random_int(183, 202),
        } as u16;
        Satellite {
            constellation: constell,
            id,
        }
    }
}

#[derive(Debug, Clone)]
pub enum Constellation {
    GPS,
    GLONASS,
    GALILEO,
    BEIDOU,
    QZSS,
}

impl Constellation {
    pub fn to_string(&self) -> String {
        match self {
            Constellation::GPS => "GPS".to_string(),
            Constellation::GLONASS => "GLONASS".to_string(),
            Constellation::GALILEO => "GALILEO".to_string(),
            Constellation::BEIDOU => "BEIDOU".to_string(),
            Constellation::QZSS => "QZSS".to_string(),
        }
    }
    pub fn to_code(&self) -> String {
        match self {
            Constellation::GPS => "GP".to_string(),
            Constellation::GLONASS => "GL".to_string(),
            Constellation::GALILEO => "GA".to_string(),
            Constellation::BEIDOU => "GB".to_string(),
            Constellation::QZSS => "GQ".to_string(),
        }
    }

    pub fn len() -> usize {
        Constellation::QZSS as usize + 1
    }

    pub fn get_random() -> Self {
        let mut rng = thread_rng();
        let range = Uniform::from(0..Constellation::len() - 1);
        let index = range.sample(&mut rng);

        match index {
            0 => Constellation::GPS,
            1 => Constellation::GLONASS,
            2 => Constellation::GALILEO,
            3 => Constellation::BEIDOU,
            _ => panic!("Invalid index"),
        }
    }
}

pub struct LocationData {
    pub latitude: String,
    pub ns: char,
    pub longitude: String,
    pub ew: char,
}

pub struct NmeaGenerator {
    rg: RandomGenerator,
}

impl NmeaGenerator {
    pub fn new() -> Self {
        NmeaGenerator {
            rg: RandomGenerator::new(),
        }
    }

    fn generate_location(&mut self) -> LocationData {
        let latitude = self.rg.random_uniform(-90.0, 90.0);
        let ns = if latitude >= 0.0 { 'N' } else { 'S' };
        let lat_deg = latitude.abs().floor();
        let lat_min = (latitude.abs() - lat_deg) * 60.0;

        let longitude = self.rg.random_uniform(-180.0, 180.0);
        let ew = if longitude >= 0.0 { 'E' } else { 'W' };
        let lon_deg = longitude.abs().floor();
        let lon_min = (longitude.abs() - lon_deg) * 60.0;

        LocationData {
            latitude: format!("{:02}{:07.4}", lat_deg, lat_min),
            ns,
            longitude: format!("{:03}{:07.4}", lon_deg, lon_min),
            ew,
        }
    }

    fn get_utc_time(&self) -> String {
        let now = Utc::now();
        now.format("%H%M%S").to_string()
    }

    fn get_utc_date(&self) -> String {
        let now = Utc::now();
        now.format("%d%m%y").to_string()
    }

    fn calculate_checksum(&self, sentence: &str) -> String {
        let mut checksum: u8 = 0;
        for c in sentence.as_bytes() {
            checksum ^= c;
        }

        format!("{:02X}", checksum)
    }

    fn complete_sentence(&self, sentence: &str) -> String {
        format!("${}*{}\r\n", sentence, self.calculate_checksum(sentence))
    }

    fn generate_gga(&mut self, loc: &LocationData, num_satellites: i32) -> String {
        let utc_time = self.get_utc_time();
        let fix_quality = self.rg.random_int(0, 5);
        let altitude = self.rg.random_uniform(0.0, 1000.0);
        let hdop = self.rg.random_uniform(0.5, 10.0);
        let geoid_height = self.rg.random_uniform(-100.0, 100.0);

        let sentence = format!(
            "GPGGA,{},{},{},{},{},{},{},{:.1},{:.1},M,{:.1},M,,",
            utc_time,
            loc.latitude,
            loc.ns,
            loc.longitude,
            loc.ew,
            fix_quality,
            num_satellites,
            hdop,
            altitude,
            geoid_height
        );

        self.complete_sentence(&sentence)
    }

    fn generate_rmc(&mut self, loc: &LocationData) -> String {
        let utc_time = self.get_utc_time();
        let status = 'A';
        let latitude = format!("{}{}", loc.latitude, loc.ns);
        let longitude = format!("{}{}", loc.longitude, loc.ew);
        let speed = self.rg.random_uniform(0.0, 100.0);
        let course = self.rg.random_uniform(0.0, 360.0);
        let utc_date = self.get_utc_date();

        let sentence = format!(
            "GPRMC,{},{},{},{},{},{:.1},{:.1},{},{},,,",
            utc_time, status, latitude, loc.ns, longitude, loc.ew, speed, course, utc_date
        );

        self.complete_sentence(&sentence)
    }

    fn generate_gll(&mut self, loc: &LocationData) -> String {
        let latitude = format!("{}{}", loc.latitude, loc.ns);
        let longitude = format!("{}{}", loc.longitude, loc.ew);
        let utc_time = self.get_utc_time();
        let status = 'A';

        let sentence = format!(
            "GPGLL,{},{},{},{},{},{}",
            latitude, loc.ns, longitude, loc.ew, utc_time, status
        );

        self.complete_sentence(&sentence)
    }

    fn generate_gsa(&mut self, satellites: &Vec<Satellite>) -> String {
        let mode = 'A';
        let fix_type = 3;
        let mut msgs = Vec::new();

        let pdop = self.rg.random_uniform(0.5, 10.0);
        let hdop = self.rg.random_uniform(0.5, 10.0);
        let vdop = self.rg.random_uniform(0.5, 10.0);
        // Separte the satellites by constellation

        let mut sats_by_constell = vec![Vec::new(); Constellation::len()];
        for sat in satellites {
            let index = match sat.constellation {
                Constellation::GPS => 0,
                Constellation::GLONASS => 1,
                Constellation::GALILEO => 2,
                Constellation::BEIDOU => 3,
                Constellation::QZSS => 4,
            };
            sats_by_constell[index].push(sat);
        }

        (0..sats_by_constell.len()).for_each(|i| {
            let constellations = &sats_by_constell[i];
            if constellations.is_empty() {
                return;
            }

            let constell = constellations[0].constellation.to_code();
            let mut sats_str = constellations
                .iter()
                .map(|sat| sat.id.to_string())
                .collect::<Vec<String>>()
                .join(",");

            for _ in constellations.len()..12 {
                sats_str.push(',');
            }

            let sentence = self.complete_sentence(
                format!("{constell}GSA,{mode},{fix_type},{sats_str},{pdop:.1},{hdop:.1},{vdop:.1}")
                    .as_str(),
            );
            msgs.push(sentence);
        });

        msgs.join("")
    }

    fn generate_gsv(&mut self, satellites: &[Satellite]) -> String {
        let num_msgs = self.rg.random_int(satellites.len() as i32, 16) as usize;
        let mut msgs = Vec::new();

        for i in 0..num_msgs {
            let start = i * 4;
            let end = if i == num_msgs - 1 {
                satellites.len()
            } else {
                (i + 1) * 4
            };
            let sats = &satellites[start..end];

            let num_sats = sats.len();
            let num_sats_str = num_sats.to_string();
            let msg_num = (i + 1).to_string();
            let total_msgs = num_msgs.to_string();

            let mut sats_str = String::new();
            for sat in sats {
                sats_str.push_str(&format!("{},{},{},", sat.id, 0, 0));
            }

            let sentence = format!(
                "GPGSV,{total_msgs},{msg_num},{num_sats_str},{sats_str}",
                total_msgs = total_msgs,
                msg_num = msg_num,
                num_sats_str = num_sats_str,
                sats_str = sats_str
            );

            msgs.push(self.complete_sentence(&sentence));
        }

        msgs.join("")
    }

    fn generate_satellites(&mut self) -> Vec<Satellite> {
        let num_satellites = self.rg.random_int(4, 12);
        let mut satellites = Vec::new();
        for _ in 0..num_satellites {
            satellites.push(Satellite::new_random());
        }

        satellites
    }

    pub fn generate_sentences(&mut self) -> String {
        let loc = self.generate_location();
        let active_satellites = self.generate_satellites();
        let num_satellites = active_satellites.len() as i32;

        let mut sentences = String::new();
        sentences.push_str(&self.generate_rmc(&loc));
        sentences.push_str(&self.generate_gga(&loc, num_satellites));
        sentences.push_str(&self.generate_gll(&loc));
        sentences.push_str(&self.generate_gsa(&active_satellites));
        sentences.push_str(&self.generate_gsv(&active_satellites));

        sentences
    }
}

File: cpp/NmeaGenerator.cpp
// NmeaGenerator.cpp
#include "NmeaGenerator.hpp"

#include <iomanip>
#include <sstream>

// Constructor
NmeaGenerator::NmeaGenerator()
    : rng_(std::random_device {}())
{
}

// Checksum calculation
std::string NmeaGenerator::calculateChecksum(const std::string& nmea_sentence) const
{
    uint8_t checksum = 0;
    for (char ch : nmea_sentence) {
        checksum ^= static_cast<uint8_t>(ch);
    }
    std::stringstream ss;
    ss << std::uppercase << std::hex << std::setw(2) << std::setfill('0') << int(checksum);
    return ss.str();
}

// Random uniform double
double NmeaGenerator::randomUniform(double min, double max)
{
    std::uniform_real_distribution<double> dist(min, max);
    return dist(rng_);
}

// Random integer
int NmeaGenerator::randomInt(int min, int max)
{
    std::uniform_int_distribution<int> dist(min, max);
    return dist(rng_);
}

// Generate current UTC time string in HHMMSS format
std::string NmeaGenerator::getUTCTime() const
{
    auto now        = std::chrono::system_clock::now();
    std::time_t t_c = std::chrono::system_clock::to_time_t(now);
    std::tm tm      = *std::gmtime(&t_c);
    char utc_time[16];
    std::strftime(utc_time, sizeof(utc_time), "%H%M%S", &tm);
    return std::string(utc_time);
}

// Generate current UTC date string in DDMMYY format
std::string NmeaGenerator::getUTCDate() const
{
    auto now        = std::chrono::system_clock::now();
    std::time_t t_c = std::chrono::system_clock::to_time_t(now);
    std::tm tm      = *std::gmtime(&t_c);
    char date_str[16];
    std::strftime(date_str, sizeof(date_str), "%d%m%y", &tm);
    return std::string(date_str);
}

// Generate shared location data
LocationData NmeaGenerator::generateLocation()
{
    LocationData loc;

    // Latitude: -90 to 90
    double latitude = randomUniform(-90.0, 90.0);
    loc.ns          = (latitude >= 0) ? 'N' : 'S';
    latitude        = std::abs(latitude);
    int lat_deg     = static_cast<int>(latitude);
    double lat_min  = (latitude - lat_deg) * 60.0;
    std::ostringstream lat_ss;
    lat_ss << std::setw(2) << std::setfill('0') << lat_deg << std::fixed << std::setprecision(4)
           << lat_min;
    loc.latitude = lat_ss.str();

    // Longitude: -180 to 180
    double longitude = randomUniform(-180.0, 180.0);
    loc.ew           = (longitude >= 0) ? 'E' : 'W';
    longitude        = std::abs(longitude);
    int lon_deg      = static_cast<int>(longitude);
    double lon_min   = (longitude - lon_deg) * 60.0;
    std::ostringstream lon_ss;
    lon_ss << std::setw(3) << std::setfill('0') << lon_deg << std::fixed << std::setprecision(4)
           << lon_min;
    loc.longitude = lon_ss.str();

    return loc;
}

// Generate satellites with different constellations
std::vector<SatelliteInfo> NmeaGenerator::generateSatellites()
{
    std::vector<SatelliteInfo> satellites;

    // GPS: PRN 1-32
    for (int i = 0; i < randomInt(4, 12); ++i) {
        satellites.push_back(SatelliteInfo { randomInt(1, 32), Constellation::GPS });
    }

    // GLONASS: PRN 65-96
    for (int i = 0; i < randomInt(4, 12); ++i) {
        satellites.push_back(SatelliteInfo { randomInt(65, 96), Constellation::GLONASS });
    }

    // Galileo: PRN 201-237
    for (int i = 0; i < randomInt(4, 12); ++i) {
        satellites.push_back(SatelliteInfo { randomInt(201, 237), Constellation::Galileo });
    }

    // Beidou: PRN 301-336
    for (int i = 0; i < randomInt(4, 12); ++i) {
        satellites.push_back(SatelliteInfo { randomInt(301, 336), Constellation::Beidou });
    }

    // QZSS: PRN 193-200
    for (int i = 0; i < randomInt(1, 4); ++i) {
        satellites.push_back(SatelliteInfo { randomInt(193, 200), Constellation::QZSS });
    }

    return satellites;
}

// Generate GPGGA sentence
std::string NmeaGenerator::generateGPGGA(const LocationData& loc, int numSatellites)
{
    std::string utc_time    = getUTCTime();
    std::string fix_quality = std::to_string(randomInt(0, 5));
    double horizontal_dil   = randomUniform(0.5, 2.5);
    double altitude         = randomUniform(10.0, 100.0);
    double geoid_sep        = randomUniform(-50.0, 50.0);

    std::ostringstream gpgga_body;
    gpgga_body << "GPGGA," << utc_time << "," << loc.latitude << "," << loc.ns << ","
               << loc.longitude << "," << loc.ew << "," << fix_quality << "," << numSatellites
               << "," << std::fixed << std::setprecision(1) << horizontal_dil << "," << altitude
               << ",M," << geoid_sep << ",M,,,";
    std::string checksum = calculateChecksum(gpgga_body.str());
    return "$" + gpgga_body.str() + "*" + checksum + "\r\n";
}

// Generate GPRMC sentence
std::string NmeaGenerator::generateGPRMC(const LocationData& loc)
{
    std::string utc_time      = getUTCTime();
    std::string date_str      = getUTCDate();

    double speed_over_ground  = randomUniform(0, 100);
    double course_over_ground = randomUniform(0, 360);

    std::ostringstream gprmc_body;
    gprmc_body << "GPRMC," << utc_time << ",A," << loc.latitude << "," << loc.ns << ","
               << loc.longitude << "," << loc.ew << "," << std::fixed << std::setprecision(1)
               << speed_over_ground << "," << course_over_ground << "," << date_str << ",,,";
    std::string checksum = calculateChecksum(gprmc_body.str());
    return "$" + gprmc_body.str() + "*" + checksum + "\r\n";
}

// Generate GPGLL sentence
std::string NmeaGenerator::generateGPGLL(const LocationData& loc)
{
    std::string utc_time = getUTCTime();

    std::ostringstream gpgll_body;
    gpgll_body << "GPGLL," << loc.latitude << "," << loc.ns << "," << loc.longitude << "," << loc.ew
               << "," << utc_time << ",A";
    std::string checksum = calculateChecksum(gpgll_body.str());
    return "$" + gpgll_body.str() + "*" + checksum + "\r\n";
}

// Generate GPGSA sentence
std::string NmeaGenerator::generateGPGSA(const std::vector<SatelliteInfo>& satellites)
{
    char mode            = 'A';
    std::string fix_type = std::to_string(randomInt(1, 3));

    std::vector<int> prn_list;
    // Select satellites that are used for the fix
    int satellites_used = randomInt(4, 12);
    for (int i = 0; i < satellites_used && i < satellites.size(); ++i) {
        prn_list.push_back(satellites[i].prn);
    }

    while (prn_list.size() < 12) {
        prn_list.push_back(0);
    }

    double pdop = randomUniform(1.0, 5.0);
    double hdop = randomUniform(1.0, 5.0);
    double vdop = randomUniform(1.0, 5.0);

    std::ostringstream gpgsa_body;
    gpgsa_body << "GPGSA," << mode << "," << fix_type;
    for (int prn : prn_list) {
        if (prn != 0) {
            gpgsa_body << "," << prn;
        } else {
            gpgsa_body << ",";
        }
    }
    gpgsa_body << "," << std::fixed << std::setprecision(1) << pdop << "," << hdop << "," << vdop;
    std::string checksum = calculateChecksum(gpgsa_body.str());
    return "$" + gpgsa_body.str() + "*" + checksum + "\r\n";
}

// Generate GxGSV sentences based on constellation
std::string NmeaGenerator::generateGxGSV(const std::vector<SatelliteInfo>& satellites, Constellation constellation)
{
    std::string message_id;
    switch (constellation) {
    case Constellation::GPS:
        message_id = "GPGSV";
        break;
    case Constellation::GLONASS:
        message_id = "GLGSV";
        break;
    case Constellation::Galileo:
        message_id = "GAGSV";
        break;
    case Constellation::Beidou:
        message_id = "GBGSV";
        break;
    case Constellation::QZSS:
        message_id = "GQZSV";
        break;
    default:
        message_id = "GPGSV";
        break;
    }

    // Filter satellites for the specified constellation
    std::vector<SatelliteInfo> constell_satellites;
    for (const auto& sat : satellites) {
        if (sat.constellation == constellation) {
            constell_satellites.push_back(sat);
        }
    }

    int total_sats = constell_satellites.size();
    if (total_sats == 0) {
        // If no satellites for this constellation, return empty string
        return "";
    }

    int sats_per_message = 4;
    int total_messages   = (total_sats + sats_per_message - 1) / sats_per_message;

    std::ostringstream gsv_sentences;

    for (int msg_num = 1; msg_num <= total_messages; ++msg_num) {
        std::ostringstream gsv_body;
        gsv_body << message_id << "," << total_messages << "," << msg_num << "," << total_sats;

        int start_idx = (msg_num - 1) * sats_per_message;
        int end_idx   = std::min(start_idx + sats_per_message, total_sats);

        for (int i = start_idx; i < end_idx; ++i) {
            int prn       = constell_satellites[i].prn;
            int elevation = randomInt(0, 90);
            int azimuth   = randomInt(0, 359);
            int snr       = randomInt(0, 50);
            gsv_body << "," << prn << "," << elevation << "," << azimuth << "," << snr;
        }

        // If less than 4 satellites in this message, fill with empty fields
        int sats_in_this_msg = end_idx - start_idx;
        for (int i = sats_in_this_msg; i < sats_per_message; ++i) {
            gsv_body << ",,,";
        }

        // Calculate checksum
        std::string checksum = calculateChecksum(gsv_body.str());

        // Complete sentence
        std::string sentence = "$" + gsv_body.str() + "*" + checksum + "\r\n";
        gsv_sentences << sentence;
    }

    return gsv_sentences.str();
}

// Generate NFIMU sentence
std::string NmeaGenerator::generateNFIMU()
{
    int calibration_status = randomInt(0, 1);
    double temperature     = randomUniform(10, 80);
    double acc_x           = randomUniform(-100, 100);
    double acc_y           = randomUniform(-100, 100);
    double acc_z           = randomUniform(-100, 100);
    double gyro_x          = randomUniform(-2 * 3.14, 2 * 3.14);
    double gyro_y          = randomUniform(-2 * 3.14, 2 * 3.14);
    double gyro_z          = randomUniform(-2 * 3.14, 2 * 3.14);

    std::string veh_acc_x, veh_acc_y, veh_acc_z;
    std::string veh_gyro_x, veh_gyro_y, veh_gyro_z;

    if (calibration_status == 1) {
        veh_acc_x  = std::to_string(acc_x + randomUniform(-10, 10));
        veh_acc_y  = std::to_string(acc_y + randomUniform(-10, 10));
        veh_acc_z  = std::to_string(acc_z + randomUniform(-10, 10));
        veh_gyro_x = std::to_string(gyro_x + randomUniform(-2 * 3.14 * 0.1, 2 * 3.14 * 0.1));
        veh_gyro_y = std::to_string(gyro_y + randomUniform(-2 * 3.14 * 0.1, 2 * 3.14 * 0.1));
        veh_gyro_z = std::to_string(gyro_z + randomUniform(-2 * 3.14 * 0.1, 2 * 3.14 * 0.1));
    } else {
        veh_acc_x = veh_acc_y = veh_acc_z = veh_gyro_x = veh_gyro_y = veh_gyro_z = "";
    }

    std::ostringstream nfimu_body;
    nfimu_body << "NFIMU," << calibration_status << "," << std::fixed << std::setprecision(4)
               << temperature << "," << acc_x << "," << acc_y << "," << acc_z << "," << gyro_x
               << "," << gyro_y << "," << gyro_z;

    // Only append veh_acc and veh_gyro if calibration_status == 1
    if (calibration_status == 1) {
        nfimu_body << "," << veh_acc_x << "," << veh_acc_y << "," << veh_acc_z << "," << veh_gyro_x
                   << "," << veh_gyro_y << "," << veh_gyro_z;
    } else {
        nfimu_body << ",,,,,"; // Placeholder commas for missing data
    }

    std::string checksum = calculateChecksum(nfimu_body.str());
    return "$" + nfimu_body.str() + "*" + checksum + "\r\n";
}

// Generate all NMEA sentences
std::string NmeaGenerator::generateAllSentences()
{
    LocationData loc                      = generateLocation();
    std::vector<SatelliteInfo> satellites = generateSatellites();

    std::ostringstream all_sentences;
    all_sentences << generateGPRMC(loc);
    all_sentences << generateGPGGA(loc, randomInt(4, 12));
    all_sentences << generateGPGSA(satellites);

    // Generate GSV sentences for each constellation
    std::vector<Constellation> constellations = {
        Constellation::GPS,
        Constellation::GLONASS,
        Constellation::Galileo,
        Constellation::Beidou,
        Constellation::QZSS
    };

    for (const auto& constell : constellations) {
        std::string gsv = generateGxGSV(satellites, constell);
        if (!gsv.empty()) {
            all_sentences << gsv;
        }
    }

    all_sentences << generateGPGLL(loc);
    all_sentences << generateNFIMU(loc);
    return all_sentences.str();
}
